core/src/test/java/org/elasticsearch/indices/store/IndicesStoreIntegrationTests.java

```diff
@@ -306,4 +307,46 @@ public boolean apply(Object o) {
         });
         return Files.exists(indexDirectory(server, index));
     }
+
+    /**
+     * This Tracer blocks cluster state processing from the moment a node starts a recovery
+     * until another node sends a shard active request.
+     * This is used to test the following:
+     * Whenever a node deletes a shard because it was relocated somewhere else, it first
+     * checks if enough other copies are started somewhere else. The node sends a ShardActiveRequest
+     * to the other nodes that should have a copy according to cluster state.
+     * The nodes that receive this request check if the shard is in state STARTED in which case they
+     * respond with "true". If they have the shard in POST_RECOVERY they register a cluster state
+     * observer that checks at each update if the shard has moved to STARTED.
+     * To test that this mechanism actually works, this can be triggered by blocking the cluster
+     * state processing when a recover starts and only unblocking it shortly after the node receives
+     * the ShardActiveRequest.
+     */
+    static class Tracer extends MockTransportService.Tracer {
+        private final SingleNodeDisruption disruption;
+        private final ESLogger logger;
+
+        Tracer(SingleNodeDisruption disruption, ESLogger logger) {
+            this.disruption = disruption;
+            this.logger = logger;
+        }
+
+        @Override
+        public void receivedRequest(long requestId, String action) {
+            if (action.equals(IndicesStore.ACTION_SHARD_EXISTS)) {
+                try {
+                    // wait a little to make sure cluster first shard active request fails. This should trigger registering a state observer.
+                    sleep(50);
```

Comment:
I'm not sure this does what you want? it blocks the thread handling the request. Maybe use latches here to signal events and control the disruption from the test? I think it will also be clearer to read.